src/server/shell/decoration/basic_decoration.h

```diff
@@ -75,25 +74,27 @@ class BasicDecoration
         std::shared_ptr<scene::Surface> const& window_surface);
     ~BasicDecoration();
 
-    void window_state_updated();
-    void input_state_updated();
+    void redraw();
 
+    void handle_input_event(std::shared_ptr<MirEvent const> const& event) override;
+    void set_scale(float scale) override;
+    void attrib_changed(mir::scene::Surface const* window_surface, MirWindowAttrib attrib, int value) override;
+    void window_resized_to(mir::scene::Surface const* window_surface, geometry::Size const& window_size) override;
+    void window_renamed(mir::scene::Surface const* window_surface, std::string const& name) override;
+
+    void input_state_changed();
     void request_move(MirInputEvent const* event);
     void request_resize(MirInputEvent const* event, MirResizeEdge edge);
     void request_toggle_maximize();
     void request_minimize();
     void request_close();
-    void set_cursor(std::string const& cursor_image_name);
 
-    void set_scale(float scale) override;
+    void set_cursor(std::string const& cursor_image_name);
 
 protected:
```

Comment:
There's no reason for `protected` here (there's rarely a good reason), this should be `private`.